akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala

```diff
@@ -243,22 +238,36 @@ object Source extends SourceApply {
    * Every connected `Sink` of this stream will see an individual stream consisting of one element.
    */
   def single[T](element: T): Source[T, Unit] =
-    apply(SingleElementPublisher(element, "SingleSource")).withAttributes(DefaultAttributes.singleSource)
+    new Source(
+      new PublisherSource(
+        SingleElementPublisher(element, "SingleSource"),
+        DefaultAttributes.singleSource,
+        shape("SingleSource")))
 
   /**
    * Create a `Source` that will continually emit the given element.
    */
   def repeat[T](element: T): Source[T, Unit] =
-    apply(new immutable.Iterable[T] {
-      override val iterator: Iterator[T] = Iterator.continually(element)
-      override def toString: String = "Iterable.continually(" + element + ")"
-    }).withAttributes(DefaultAttributes.repeat)
+    new Source(
+      new PublisherSource(
+        SingleElementPublisher(
+          new immutable.Iterable[T] {
+            override val iterator: Iterator[T] = Iterator.continually(element)
+            override def toString: String = "repeat(" + element + ")"
+          }, "RepeatSource"),
+        DefaultAttributes.repeat,
+        shape("RepeatSource"))).mapConcat(id)
 
   /**
    * A `Source` with no elements, i.e. an empty stream that is completed immediately for every connected `Sink`.
    */
   def empty[T]: Source[T, Unit] = _empty
-  private[this] val _empty: Source[Nothing, Unit] = apply(EmptyPublisher).withAttributes(DefaultAttributes.emptySource)
+  private[this] val _empty: Source[Nothing, Unit] =
+    new Source(
+      new PublisherSource[Nothing](
+        EmptyPublisher,
+        DefaultAttributes.emptySource,
+        shape("EmptySource")))
```

Comment:
There's quite a bit of repetition here when creating these sources. Is there any reason why the PublisherSource can't create its own SourceShape given that it is handed the name anyway?